[gradle] Support for multi-project builds

<!-- Thanks for filing an issue! Please use this template to provide more context for the issue . -->

I think a part of this would be also thinking about the layering - I guess project dependencies should go in a separate layer on top of upstream dependencies, or maybe even together with the app.

**Description of the issue**:

Repro: https://github.com/anuraaga/gradle-jib-multiproject-example

When trying to run `jibExportDockerContext` for a gradle project with multiple subprojects that depend on each other, I found that the project dependency only has `compileJava` run, not `jar`. Because of this, `jibExportDockerContext` fails when trying to copy the jar with 

```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':server:jibExportDockerContext'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.execution.taskgraph.LocalTaskInfoExecutor.execute(LocalTaskInfoExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:273)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareWorkItemExecutor.execute(DefaultTaskExecutionGraph.java:258)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:135)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.execute(DefaultTaskPlanExecutor.java:200)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:191)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.GradleException: Export Docker context failed, perhaps you should check if the command-line option `--jibTargetDir` is set correctly
        at com.google.cloud.tools.jib.gradle.DockerContextTask.generateDockerContext(DockerContextTask.java:144)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:786)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:753)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
        ... 31 more
Caused by: java.nio.file.NoSuchFileException: C:\tools\msys64\home\Anuraag\git\gradle-jib-multiproject-example\lib\build\libs\lib.jar
        at com.google.cloud.tools.jib.filesystem.FileOperations.lambda$copy$0(FileOperations.java:43)
        at com.google.cloud.tools.jib.filesystem.FileOperations.copy(FileOperations.java:50)
        at com.google.cloud.tools.jib.docker.DockerContextGenerator.generate(DockerContextGenerator.java:209)
        at com.google.cloud.tools.jib.gradle.DockerContextTask.generateDockerContext(DockerContextTask.java:128)
        ... 48 more
```

**Expected behavior**:

`jar` task should be run for project dependencies

**Steps to reproduce**: <!-- Please provide a minimal and precise series of steps -->

In a multi-project build, where `server` depends on `lib`, run

```
$ git clone https://github.com/anuraaga/gradle-jib-multiproject-example
$ cd gradle-jib-multiproject-example
$ ./gradlew :server:jibExportDockerContext --stacktrace --info
```

There will be a `NoSuchFileException` for `lib.jar`

**Environment**: <!-- OS, Maven, and other relevant environment information -->

Windows 10+ Msys2, Gradle 4.9, Java 10

**`jib-gradle-plugin` Configuration**: <!-- Delete this section if not used -->

Adapted:
```groovy
jib {
  from {
    image = 'openjdk:10-jre-slim'
  }
  to {
    image = 'gcr.io/test-project/test-image'
  }
  container {
    mainClass = 'Main'
  }
}
```

**Log output**: <!-- If applicable, provide relevant log output -->
```

```
**Additional Information**: <!-- Any additional information that may be helpful -->


<!-- Thanks for contributing! -->

--- a/jib-gradle-plugin/CHANGELOG.md
+++ b/jib-gradle-plugin/CHANGELOG.md
@@ -11,6 +11,8 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Gradle project dependencies have their `assemble` task run before running a jib task ([#815](https://github.com/GoogleContainerTools/jib/issues/815))
+
 ## 0.9.8
 
 ### Added

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java
@@ -40,6 +40,42 @@ public class JibPlugin implements Plugin<Project> {
   @VisibleForTesting static final String BUILD_DOCKER_TASK_NAME = "jibDockerBuild";
   @VisibleForTesting static final String DOCKER_CONTEXT_TASK_NAME = "jibExportDockerContext";
 
+  private static void checkGradleVersion() {
+    if (GRADLE_MIN_VERSION.compareTo(GradleVersion.current()) > 0) {
+      throw new GradleException(
+          "Detected "
+              + GradleVersion.current()
+              + ", but jib requires "
+              + GRADLE_MIN_VERSION
+              + " or higher. You can upgrade by running 'gradle wrapper --gradle-version="
+              + GRADLE_MIN_VERSION.getVersion()
+              + "'.");
+    }
+  }
+
+  /**
+   * Collects all assemble tasks for project dependencies of the style "compile project(':mylib')"
+   * for any kind of configuration [compile, runtime, etc]. It potentially will collect common test
+   * libraries in configs like [test, integrationTest, etc], but it's either that or filter based on
+   * a configuration containing the word "test" which feels dangerous.
+   *
+   * @param project this project we are containerizing
+   * @return a list of "assemble" tasks associated with projects that this project depends on.
+   */
+  @VisibleForTesting
+  static List<Task> getProjectDependencyAssembleTasks(Project project) {
+    return project
+        .getConfigurations()
+        .stream()
+        .map(Configuration::getDependencies)
+        .flatMap(DependencySet::stream)
+        .filter(ProjectDependency.class::isInstance)
+        .map(ProjectDependency.class::cast)
+        .map(ProjectDependency::getDependencyProject)
+        .map(subProject -> subProject.getTasks().getByPath(BasePlugin.ASSEMBLE_TASK_NAME))
+        .collect(Collectors.toList());
+  }
+
   @Override
   public void apply(Project project) {
     checkGradleVersion();
@@ -68,18 +104,23 @@ public class JibPlugin implements Plugin<Project> {
             .create(BUILD_TAR_TASK_NAME, BuildTarTask.class)
             .setJibExtension(jibExtension);
 
-    // Has all tasks depend on the 'classes' task.
     project.afterEvaluate(
         projectAfterEvaluation -> {
           try {
-            List<Task> taskDependencies = getProjectDependencyAssembleTasks(projectAfterEvaluation);
+            // Find project dependencies
+            List<Task> computedDependencies =
+                getProjectDependencyAssembleTasks(projectAfterEvaluation);
+            // Has all tasks depend on the 'classes' task.
             Task classesTask = projectAfterEvaluation.getTasks().getByPath("classes");
-            taskDependencies.add(classesTask);
+            computedDependencies.add(classesTask);
 
-            buildImageTask.dependsOn(taskDependencies.toArray());
-            dockerContextTask.dependsOn(taskDependencies.toArray());
-            buildDockerTask.dependsOn(taskDependencies.toArray());
-            buildTarTask.dependsOn(taskDependencies.toArray());
+            // dependsOn takes an Object... type
+            Object[] dependenciesArray = computedDependencies.toArray();
+
+            buildImageTask.dependsOn(dependenciesArray);
+            dockerContextTask.dependsOn(dependenciesArray);
+            buildDockerTask.dependsOn(dependenciesArray);
+            buildTarTask.dependsOn(dependenciesArray);
 
           } catch (UnknownTaskException ex) {
             throw new GradleException(
@@ -89,40 +130,5 @@ public class JibPlugin implements Plugin<Project> {
                 ex);
           }
         });
-  }
-
-  /**
-   * Collects all assemble tasks for project dependencies of the style "compile project(':mylib')"
-   * for any kind of configuration [compile, runtime, etc]. It potentially will collect common test
-   * libraries in configs like [test, integrationTest, etc], but it's either that or filter based on
-   * a configuration containing the word "test" which feels dangerous.
-   *
-   * @param project this project we are containerizing
-   * @return a list of "assemble" tasks associated with projects that this project depends on.
-   */
-  static List<Task> getProjectDependencyAssembleTasks(Project project) {
-    return project
-        .getConfigurations()
-        .stream()
-        .map(Configuration::getDependencies)
-        .flatMap(DependencySet::stream)
-        .filter(ProjectDependency.class::isInstance)
-        .map(ProjectDependency.class::cast)
-        .map(ProjectDependency::getDependencyProject)
-        .map(subProject -> subProject.getTasks().getByPath(BasePlugin.ASSEMBLE_TASK_NAME))
-        .collect(Collectors.toList());
-  }
-
-  private static void checkGradleVersion() {
-    if (GRADLE_MIN_VERSION.compareTo(GradleVersion.current()) > 0) {
-      throw new GradleException(
-          "Detected "
-              + GradleVersion.current()
-              + ", but jib requires "
-              + GRADLE_MIN_VERSION
-              + " or higher. You can upgrade by running 'gradle wrapper --gradle-version="
-              + GRADLE_MIN_VERSION.getVersion()
-              + "'.");
-    }
   }
 }

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java
@@ -17,11 +17,17 @@
 package com.google.cloud.tools.jib.gradle;
 
 import com.google.common.annotations.VisibleForTesting;
+import java.util.List;
+import java.util.stream.Collectors;
 import org.gradle.api.GradleException;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
 import org.gradle.api.UnknownTaskException;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.DependencySet;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.plugins.BasePlugin;
 import org.gradle.util.GradleVersion;
 
 public class JibPlugin implements Plugin<Project> {
@@ -66,12 +72,14 @@ public class JibPlugin implements Plugin<Project> {
     project.afterEvaluate(
         projectAfterEvaluation -> {
           try {
+            List<Task> taskDependencies = getProjectDependencyAssembleTasks(projectAfterEvaluation);
             Task classesTask = projectAfterEvaluation.getTasks().getByPath("classes");
+            taskDependencies.add(classesTask);
 
-            buildImageTask.dependsOn(classesTask);
-            dockerContextTask.dependsOn(classesTask);
-            buildDockerTask.dependsOn(classesTask);
-            buildTarTask.dependsOn(classesTask);
+            buildImageTask.dependsOn(taskDependencies.toArray());
+            dockerContextTask.dependsOn(taskDependencies.toArray());
+            buildDockerTask.dependsOn(taskDependencies.toArray());
+            buildTarTask.dependsOn(taskDependencies.toArray());
 
           } catch (UnknownTaskException ex) {
             throw new GradleException(
@@ -81,6 +89,28 @@ public class JibPlugin implements Plugin<Project> {
                 ex);
           }
         });
+  }
+
+  /**
+   * Collects all assemble tasks for project dependencies of the style "compile project(':mylib')"
+   * for any kind of configuration [compile, runtime, etc]. It potentially will collect common test
+   * libraries in configs like [test, integrationTest, etc], but it's either that or filter based on
+   * a configuration containing the word "test" which feels dangerous.
+   *
+   * @param project this project we are containerizing
+   * @return a list of "assemble" tasks associated with projects that this project depends on.
+   */
+  static List<Task> getProjectDependencyAssembleTasks(Project project) {
+    return project
+        .getConfigurations()
+        .stream()
+        .map(Configuration::getDependencies)
+        .flatMap(DependencySet::stream)
+        .filter(ProjectDependency.class::isInstance)
+        .map(ProjectDependency.class::cast)
+        .map(ProjectDependency::getDependencyProject)
+        .map(subProject -> subProject.getTasks().getByPath(BasePlugin.ASSEMBLE_TASK_NAME))
+        .collect(Collectors.toList());
   }
 
   private static void checkGradleVersion() {

--- a/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java
+++ b/jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java
@@ -40,19 +40,6 @@ public class JibPlugin implements Plugin<Project> {
   @VisibleForTesting static final String BUILD_DOCKER_TASK_NAME = "jibDockerBuild";
   @VisibleForTesting static final String DOCKER_CONTEXT_TASK_NAME = "jibExportDockerContext";
 
-  private static void checkGradleVersion() {
-    if (GRADLE_MIN_VERSION.compareTo(GradleVersion.current()) > 0) {
-      throw new GradleException(
-          "Detected "
-              + GradleVersion.current()
-              + ", but jib requires "
-              + GRADLE_MIN_VERSION
-              + " or higher. You can upgrade by running 'gradle wrapper --gradle-version="
-              + GRADLE_MIN_VERSION.getVersion()
-              + "'.");
-    }
-  }
-
   /**
    * Collects all assemble tasks for project dependencies of the style "compile project(':mylib')"
    * for any kind of configuration [compile, runtime, etc]. It potentially will collect common test
@@ -74,6 +61,19 @@ public class JibPlugin implements Plugin<Project> {
         .map(ProjectDependency::getDependencyProject)
         .map(subProject -> subProject.getTasks().getByPath(BasePlugin.ASSEMBLE_TASK_NAME))
         .collect(Collectors.toList());
+  }
+
+  private static void checkGradleVersion() {
+    if (GRADLE_MIN_VERSION.compareTo(GradleVersion.current()) > 0) {
+      throw new GradleException(
+          "Detected "
+              + GradleVersion.current()
+              + ", but jib requires "
+              + GRADLE_MIN_VERSION
+              + " or higher. You can upgrade by running 'gradle wrapper --gradle-version="
+              + GRADLE_MIN_VERSION.getVersion()
+              + "'.");
+    }
   }
 
   @Override

--- a/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java
+++ b/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java
@@ -16,14 +16,15 @@
 
 package com.google.cloud.tools.jib.gradle;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.Arrays;
 import java.util.Collections;
+import java.util.List;
 import java.util.stream.Collectors;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
@@ -39,6 +40,12 @@ import org.junit.rules.TemporaryFolder;
 /** Tests for {@link JibPlugin}. */
 public class JibPluginTest {
 
+  private static final List<String> KNOWN_JIB_TASKS =
+      ImmutableList.of(
+          JibPlugin.BUILD_IMAGE_TASK_NAME,
+          JibPlugin.BUILD_DOCKER_TASK_NAME,
+          JibPlugin.DOCKER_CONTEXT_TASK_NAME,
+          JibPlugin.BUILD_TAR_TASK_NAME);
   @Rule public TemporaryFolder testProjectRoot = new TemporaryFolder();
 
   @Test
@@ -120,25 +127,27 @@ public class JibPluginTest {
 
     // check by applying the jib plugin and inspect the task dependencies
     rootProject.getPluginManager().apply("com.google.cloud.tools.jib");
+
+    // add a custom task that our jib tasks depend on to ensure we do not overwrite this dependsOn
+    Task dependencyTask =
+        rootProject.getTasks().create("myCustomTask", task -> System.out.println("test"));
+    KNOWN_JIB_TASKS.forEach(
+        taskName -> rootProject.getTasks().getByPath(taskName).dependsOn(dependencyTask));
+
     ((ProjectInternal) rootProject).evaluate();
 
-    Arrays.asList(
-            JibPlugin.BUILD_IMAGE_TASK_NAME,
-            JibPlugin.DOCKER_CONTEXT_TASK_NAME,
-            JibPlugin.BUILD_DOCKER_TASK_NAME,
-            JibPlugin.BUILD_TAR_TASK_NAME)
-        .forEach(
-            taskName -> {
-              Assert.assertEquals(
-                  ImmutableSet.of(":sub:assemble", ":classes"),
-                  rootProject
-                      .getTasks()
-                      .getByPath(taskName)
-                      .getDependsOn()
-                      .stream()
-                      .map(Task.class::cast)
-                      .map(Task::getPath)
-                      .collect(Collectors.toSet()));
-            });
+    KNOWN_JIB_TASKS.forEach(
+        taskName -> {
+          Assert.assertEquals(
+              ImmutableSet.of(":sub:assemble", ":classes", ":myCustomTask"),
+              rootProject
+                  .getTasks()
+                  .getByPath(taskName)
+                  .getDependsOn()
+                  .stream()
+                  .map(Task.class::cast)
+                  .map(Task::getPath)
+                  .collect(Collectors.toSet()));
+        });
   }
 }

--- a/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java
+++ b/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java
@@ -16,10 +16,19 @@
 
 package com.google.cloud.tools.jib.gradle;
 
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.stream.Collectors;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.testfixtures.ProjectBuilder;
 import org.gradle.testkit.runner.GradleRunner;
 import org.gradle.testkit.runner.UnexpectedBuildFailure;
 import org.junit.Assert;
@@ -67,5 +76,69 @@ public class JibPluginTest {
                       + JibPlugin.GRADLE_MIN_VERSION.getVersion()
                       + "'."));
     }
+  }
+
+  @Test
+  public void testGetProjectDependencyAssembleTasks() {
+    // root project is our jib packaged service
+    Project rootProject =
+        ProjectBuilder.builder().withProjectDir(testProjectRoot.getRoot()).withName("root").build();
+    rootProject.getPluginManager().apply("java");
+
+    // our service DOES depend on this, and jib should trigger an assemble from this project
+    Project subProject =
+        ProjectBuilder.builder()
+            .withParent(rootProject)
+            .withProjectDir(testProjectRoot.getRoot())
+            .withName("sub")
+            .build();
+    subProject.getPluginManager().apply("java");
+
+    // our service doesn't depend on this, and jib should NOT trigger an assemble from this project
+    Project unrelatedSubProject =
+        ProjectBuilder.builder()
+            .withParent(rootProject)
+            .withProjectDir(testProjectRoot.getRoot())
+            .withName("unrelated")
+            .build();
+    unrelatedSubProject.getPluginManager().apply("java");
+
+    // equivalent of "compile project(':sub')" on the root(jib) project
+    rootProject
+        .getConfigurations()
+        .getByName("compile")
+        .getDependencies()
+        .add(rootProject.getDependencies().project(ImmutableMap.of("path", subProject.getPath())));
+
+    // programmatic check
+    Assert.assertEquals(
+        Collections.singletonList(":sub:assemble"),
+        JibPlugin.getProjectDependencyAssembleTasks(rootProject)
+            .stream()
+            .map(Task::getPath)
+            .collect(Collectors.toList()));
+
+    // check by applying the jib plugin and inspect the task dependencies
+    rootProject.getPluginManager().apply("com.google.cloud.tools.jib");
+    ((ProjectInternal) rootProject).evaluate();
+
+    Arrays.asList(
+            JibPlugin.BUILD_IMAGE_TASK_NAME,
+            JibPlugin.DOCKER_CONTEXT_TASK_NAME,
+            JibPlugin.BUILD_DOCKER_TASK_NAME,
+            JibPlugin.BUILD_TAR_TASK_NAME)
+        .forEach(
+            taskName -> {
+              Assert.assertEquals(
+                  ImmutableSet.of(":sub:assemble", ":classes"),
+                  rootProject
+                      .getTasks()
+                      .getByPath(taskName)
+                      .getDependsOn()
+                      .stream()
+                      .map(Task.class::cast)
+                      .map(Task::getPath)
+                      .collect(Collectors.toSet()));
+            });
   }
 }

--- a/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java
+++ b/jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java
@@ -24,7 +24,6 @@ import java.io.InputStream;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Collections;
-import java.util.List;
 import java.util.stream.Collectors;
 import org.gradle.api.Project;
 import org.gradle.api.Task;
@@ -40,7 +39,7 @@ import org.junit.rules.TemporaryFolder;
 /** Tests for {@link JibPlugin}. */
 public class JibPluginTest {
 
-  private static final List<String> KNOWN_JIB_TASKS =
+  private static final ImmutableList<String> KNOWN_JIB_TASKS =
       ImmutableList.of(
           JibPlugin.BUILD_IMAGE_TASK_NAME,
           JibPlugin.BUILD_DOCKER_TASK_NAME,
@@ -129,8 +128,7 @@ public class JibPluginTest {
     rootProject.getPluginManager().apply("com.google.cloud.tools.jib");
 
     // add a custom task that our jib tasks depend on to ensure we do not overwrite this dependsOn
-    Task dependencyTask =
-        rootProject.getTasks().create("myCustomTask", task -> System.out.println("test"));
+    Task dependencyTask = rootProject.getTasks().create("myCustomTask", task -> {});
     KNOWN_JIB_TASKS.forEach(
         taskName -> rootProject.getTasks().getByPath(taskName).dependsOn(dependencyTask));
 

